### ISSUE TITLE
` Layout/ClassStructure` syntax error / cloberring error auto-corrections

### ISSUE THREAD

**Issue by marcandre** [rubocop/rubocop#9723]
` Layout/ClassStructure` has a bad algorithm for moving code with respect to empty lines before/after. Look for `+1` in the code.

### Heredocs

Looks like the cop assumes that heredocs are always followed by a empty line

```ruby
      class Foo
        attr_reader :foo
        CONSTANT = <<~EOS
        ^^^^^^^^^^^^^^^^^ `constants` is supposed to appear [...]
          abc
        EOS
      end
```

gets auto-corrected to:
```ruby
      class Foo
        CONSTANT = <<~EOS
          abc
        EOS
        attr_reader :fooend
```

If the config's order is switched, then
```ruby
        class Foo
          CONSTANT = <<~EOS
            #{str}
          EOS
          attr_reader :foo
          ^^^^^^^^^^^^^ [...]
        end
```

generates a cloberring error.

### Defs

Defs (except `initialize`??) can create issues when not separated with empty lines. E.g.:

```ruby
        class Foo
          private
          def foo
          end
          public
          def bar
          ^^^^^^^ [...]
          end
        end
 ```
gets auto-corrected to:
```ruby
        class Foo
          public
          def bar
          ^^^^^^^ [...]
          end
          private
          def foo
          endend
```

In general, it should have been clear that to swap two blocks of code and keep good empty lines is not possible with a single "cut" and "paste", but that is what the current code is attempting.

Noticed in my refactor, I'll handle it.

**Comment by bbatsov** [comment]
Closing as stale — zero engagement in nearly 5 years. The other Layout/ClassStructure issues (#9561, #9574, #9722, #9738, #9739) cover related problems and remain open.
